src/args.rs

```diff
@@ -107,10 +119,137 @@ pub struct Args {
+    #[arg(long, value_name = "WHEN"/*, default_value_t = ColorChoiceArg::Auto*/)]
+    pub color: Option<ColorChoiceArg>,
+
+    /// This is a convenience alias for '--color always --heading
+    /// --line-number'.
+    ///
+    /// This flag is useful when you still want pretty output even if you're
+    /// piping tree-sitter-grep to another program or file. For example:
+    /// 'tree-sitter-grep -p -q "(function_item) @c" | less -R'.
+    #[arg(short = 'p', long)]
+    pub pretty: bool,
+
+    /// This flag prints the file path above clusters of matches from each file
+    /// instead of printing the file path as a prefix for each matched line.
+    ///
+    /// This is the default mode when printing to a terminal.
+    ///
+    /// This overrides the --no-heading flag.
+    #[arg(long)]
+    pub heading: bool,
+
+    /// Don't group matches by each file.
+    ///
+    /// If --no-heading is provided in addition to the -H/--with-filename flag,
+    /// then file paths will be printed as a prefix for every matched line.
+    /// This is the default mode when not printing to a terminal.
+    ///
+    /// This overrides the --heading flag.
+    #[arg(long, overrides_with = "heading")]
+    pub no_heading: bool,
+
+    /// Display the file path for matches.
+    ///
+    /// This is the default when more than one file is searched. If --heading is
+    /// enabled (the default when printing to a terminal), the file path will be
+    /// shown above clusters of matches from each file; otherwise, the file name
+    /// will be shown as a prefix for each matched line.
+    ///
+    /// This flag overrides --no-filename.
+    #[arg(short = 'H', long)]
+    pub with_filename: bool,
+
+    /// Never print the file path with the matched lines.
+    ///
+    /// This is the default when tree-sitter-grep is explicitly instructed to
+    /// search one file or stdin.
+    ///
+    /// This flag overrides --with-filename.
+    #[arg(short = 'I', long, overrides_with = "with_filename")]
+    pub no_filename: bool,
+
+    /// Show line numbers (1-based).
+    ///
+    /// This is enabled by default when searching in a terminal.
+    #[arg(short = 'n', long)]
+    pub line_number: bool,
+
+    /// Suppress line numbers.
+    ///
+    /// This is enabled by default when not searching in a terminal.
+    #[arg(short = 'N', long, overrides_with = "line_number")]
+    pub no_line_number: bool,
+
+    /// Show column numbers (1-based).
+    ///
+    /// This only shows the column numbers for the first match on each line.
+    /// This does not try to account for Unicode. One byte is equal to one
+    /// column. This implies --line-number.
+    ///
+    /// This flag can be disabled with --no-column.
+    #[arg(long)]
+    pub column: bool,
+
+    #[arg(long, hide = true, overrides_with = "column")]
+    pub no_column: bool,
 }
 
 impl Args {
-    fn use_paths(&self) -> Vec<PathBuf> {
+    pub fn use_paths(&self) -> Vec<PathBuf> {
```

Comment:
You could probably write a macro which does the once-cell thing that you want to the function it annotates?

Reply:
Ah yeah, I didn't really think about how it'd have to be per-instance.

That seems like a reasonable approach to me (the once cell field). I wonder if clap exposes some option to call a constructor for `Args` so that you could at startup calculate this stuff rather than having to do it lazily. (It's not like it's not going to be used, so having to do work to make it smartly lazy seems kind of silly.)